linking.go

```diff
@@ -2,147 +2,138 @@ package ipld
-	// NodeAssembler accumulates the final results (which you can
-	// presumably access from elsewhere; Load is designed not to know
-	// about this).
-	Load(context.Context, LinkContext, NodeAssembler, Loader) error
+// This file contains all the functions on LinkSystem.
+// These are the helpful, user-facing functions we expect folks to use "most of the time" when loading and storing data.
 
-	// LinkBuilder returns a handle to any parameters of the Link which
-	// are needed to create a new Link of the same style but with new content.
-	// (It's much like the relationship of Node/NodeBuilder.)
-	//
-	// (If you're familiar with CIDs, you can think of this method as
-	// corresponding closely to `cid.Prefix()`, just more abstractly.)
-	LinkBuilder() LinkBuilder
+// Varations:
+// - Load vs Store vs ComputeLink
+// - With or without LinkContext?
+//   - Brevity would be nice but I can't think of what to name the functions, so: everything takes LinkContext.  Zero value is fine though.
+// - [for load direction only]: Prototype (and return Node|error) or Assembler (and just return error)?
+//   - naming: Load vs Fill.
+// - 'Must' variants.
 
-	// String should return a reasonably human-readable debug-friendly
-	// representation of a Link.  It should only be used for debug and
-	// log message purposes; there is no contract that requires that the
-	// string be able to be parsed back into a reified Link.
-	String() string
-}
+// Can we get as far as a `QuickLoad(lnk Link) (Node, error)` function, which doesn't even ask you for a NodePrototype?
+//  No, not quite.  (Alas.)  If we tried to do so, and make it use `basicnode.Prototype`, we'd have import cycles; ded.
+
+// other notes for the commit log & PR:
+//  - curently ditched the BytesAccessor hack.
+//    - not super clear if that'll still be a bonus, given that we de-kinked a double-buffering around hashers by just ignoring a multihash design flaw in this rewrite.
+//    - always been a tad uncomfortable with that hack anyway (causes the reader to not advance: weird abuse of semantic contract, and hard/non-obvious to opt-out of).
 
-// LinkBuilder encapsulates any implementation details and parameters
-// necessary for taking a Node and converting it to a serial representation
-// and returning a Link to that data.
-//
-// The serialized bytes will be routed through the provided Storer system,
-// which is expected to store them in some way such that a related Loader
-// system can later use the Link and an associated Loader to load nodes
-// of identical content.
-//
-// LinkBuilder, like Link, is an abstract interface.
-// If using CIDs as an implementation, LinkBuilder will encapsulate things
-// like multihashType, multicodecType, and cidVersion, for example.
-type LinkBuilder interface {
-	Build(context.Context, LinkContext, Node, Storer) (Link, error)
+func (ls *LinkSystem) Load(lnkCtx LinkContext, lnk Link, np NodePrototype) (Node, error) {
```

Comment:
one design thing this setup leaves open that is a bit worrying is that as it gets used for applied subsets of IPLD dags, we'll end up with implementations that don't fully fill out LinkContext, and will be fine for their initial use case, but will then end up in designs where tools aren't generally applicable to IPLD data because of that specificity.

Reply:
I'm not super worried?  The documentation does make an explicit statement that a zero-value LinkContext is to be considered acceptable.  Also, that values in a LinkContext should only ever affect data availability, not actual content.

I can't enforce this programmatically, of course; that would require some kind of magic.  But: that's sort of inevitable.  Even if we took out LinkContext entirely, we'd still need a context.Context here... and that still has the arbitrary attachment feature.  (Which is actually desired.  I fully expect someone to use this to tunnel auth credentials, at some point.)

I don't really expect people to have a strong incentive to do unlawful/inadvisable things here, either.

(Although... speaking of "only data availability effects"... that does remind me that we need to get a _lot_ more buckled down on well-typed errors throughout this whole library.)